gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/converters/EntityServiceAndGatewayServiceConverter.java

```diff
@@ -205,6 +197,34 @@ private static Function convertToQueryFunction(
         {
           builder.setFunctionName(function.getFunction().name()).setAlias(function.getAlias());
 
+          // Backward compatibility to handle long values
+          if (function.getFunction() == FunctionType.AVGRATE) {
+            List<org.hypertrace.gateway.service.v1.common.Expression> columns =
+                function.getArgumentsList().stream()
+                    .filter(
+                        e ->
+                            e.hasLiteral()
+                                && e.getLiteral().getValue().getValueType()
+                                    == org.hypertrace.gateway.service.v1.common.ValueType.LONG)
```

Comment:
So as I'm currently reading this, we're special casing long args for AVG_RATE and converting them to ISO string, but if it's not a avg rate + long, it hits the default case?

The logic here is really confusing, you never want to have nested break statements inside conditions - very error prone.

What I would recommend is to maintain the avg rate case, and then pass the function expression into a normalizer that handles taking either the long or iso version and standardizing them into just the iso version. Same util can be used at validation time. Then, it can come through the standard conversion logic.

Reply:
resolved